OnReceived

mikaelm12 opened the pull request:
Removed the Receive event in favor of an OnReceived method as detailed in https://github.com/aspnet/SignalR/issues/900

Roast Me

dnfclas commented:
@mikaelm12,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -62,7 +63,7 @@ public HubConnection(IConnection connection, IHubProtocol protocol, ILoggerFacto
             _loggerFactory = loggerFactory ?? NullLoggerFactory.Instance;
             _logger = _loggerFactory.CreateLogger<HubConnection>();
-            _connection.Received += OnDataReceivedAsync;
+            _connection.OnReceived((data, state) => OnDataReceivedAsync(data), null);
```
Is `this` captured in the closure? If so you could pass `this` as the state and then cast it to `HubConnection` to call `OnDataReceivedAsync`

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -275,15 +287,19 @@ private Task Shutdown(Exception ex = null)
             }
-
+            var tasks = new List<Task>();
+            foreach(var handler in handlers)
```
nit: formatting

moozzyk replied:
Isn't it super racey? The list could be modified if someone happens to be adding or removing a handler while a message is received.

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -333,6 +349,7 @@ private void ThrowIfConnectionTerminated(string invocationId)
         }
 
         private string GetNextId() => Interlocked.Increment(ref _nextId).ToString();
+        private string GetNextCallBackId() => Interlocked.Increment(ref _nextCallBackId).ToString();
```
What is this?

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -377,6 +394,21 @@ private bool TryRemoveInvocation(string invocationId, out InvocationRequest irq)
 
+        private class Subscription : IDisposable
+        {
+            private InvocationHandler _handler;
```
readonly

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -398,24 +430,30 @@ public Type GetReturnType(string invocationId)
                     return Type.EmptyTypes;
                 }
-                return handler.ParameterTypes;
+                if (handlers.Count > 0)
```
Write a comment about the fact that we're using the first handler to determine the parameter types.

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnectionExtensions.cs:
```diff
@@ -10,69 +10,70 @@ namespace Microsoft.AspNetCore.SignalR.Client
+        private static IDisposable On(this HubConnection hubConnetion, string methodName, Type[] parameterTypes, Action<object[]> handler)
         {
-            hubConnetion.On(methodName, parameterTypes, (parameters) =>
+            //string methodName, Type[] parameterTypes, Func<object[], object, Task> handler, object state)
```
Dead code?

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnectionExtensions.cs:
```diff
@@ -10,69 +10,70 @@ namespace Microsoft.AspNetCore.SignalR.Client
                 handler(parameters);
                 return Task.CompletedTask;
-            });
+            }, null);
```
You can avoid the closure now and pass the handler as state.

davidfowl commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +440,51 @@ private void SetTransferMode(TransferMode transferMode)
             _httpClient.Dispose();
         }
 
+        private int GetNextCallBackId() => Interlocked.Increment(ref _callBackId);
```
Why?

davidfowl commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -40,12 +42,13 @@ public class HttpConnection : IConnection
 
         private ReadableChannel<byte[]> Input => _transportChannel.In;
         private WritableChannel<SendMessage> Output => _transportChannel.Out;
+        private List<ReceiveCallBack> _callBacks = new List<ReceiveCallBack>();
```
readonly

davidfowl replied:
nit: style these should be with the other fields.

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -275,15 +287,19 @@ private Task Shutdown(Exception ex = null)
             // TODO: Return values
             // TODO: Dispatch to a sync context to ensure we aren't blocking this loop.
-            return handler.Handler(invocation.Arguments);
+            return Task.WhenAll(tasks);
```
Lets avoid calling these in parallel. Call each one sequentially.

moozzyk replied:
What happens if a user throws?

davidfowl commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +440,51 @@ private void SetTransferMode(TransferMode transferMode)
+
+        private class ReceiveCallBack
+        {
+            public int CallBackID { get; }
```
This isn't needed.

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -398,24 +430,30 @@ public Type GetReturnType(string invocationId)
+                {
+                    return handlers[0].ParameterTypes;
+                }
+                throw new FormatException($"There are no callback registered for the method '{methodName}'");
```
"callbacks"

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -275,15 +287,19 @@ private Task Shutdown(Exception ex = null)
+            var tasks = new List<Task>();
+            foreach(var handler in handlers)
+            {
+                tasks.Add(handler.Handler(invocation.Arguments, null));
```
null seems wrong, it should be `handler.State`?

BrennanConroy replied:
Which also means you should add some tests using state

BrennanConroy commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -338,17 +341,10 @@ private void SetTransferMode(TransferMode transferMode)
 
-                            var receivedHandler = Received;
-                            if (receivedHandler != null)
+                            //First item in the tuple is the callback, second item is the state object.
```
Remove

davidfowl commented on samples/ClientSample/RawSample.cs:
```diff
@@ -40,7 +40,7 @@ internal static void Register(CommandLineApplication app)
             {
                 var cts = new CancellationTokenSource();
-                connection.Received += data => Console.Out.WriteLineAsync($"{Encoding.UTF8.GetString(data)}");
+                connection.OnReceived((data, state) => Console.Out.WriteLineAsync($"{Encoding.UTF8.GetString(data)}"), null);
```
Add an extension method that doesn't require state

BrennanConroy commented:
Agreed

moozzyk commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -398,24 +430,30 @@ public Type GetReturnType(string invocationId)
 
             public Type[] GetParameterTypes(string methodName)
```
probably needs a lock - if the last subscription for the method happens to be removed on a different thread accessing the handler will fail.

moozzyk commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -398,24 +430,30 @@ public Type GetReturnType(string invocationId)
+            public object State { get; }
 
-            public InvocationHandler(Type[] parameterTypes, Func<object[], Task> handler)
+            public InvocationHandler(Type[] parameterTypes, Func<object[], object, Task> handler, object state)
```
nit: I would rename Handler/handler to Callback/callback

moozzyk commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -40,12 +42,13 @@ public class HttpConnection : IConnection
 
         private ReadableChannel<byte[]> Input => _transportChannel.In;
         private WritableChannel<SendMessage> Output => _transportChannel.Out;
+        private List<ReceiveCallBack> _callBacks = new List<ReceiveCallBack>();
```
What if we just had just one callback (a Func) on HttpConnection? Just wondering when/if multiple callbacks are useful at this level...

moozzyk commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -338,17 +341,10 @@ private void SetTransferMode(TransferMode transferMode)
-                            var receivedHandler = Received;
-                            if (receivedHandler != null)
+                            //First item in the tuple is the callback, second item is the state object.
+                            foreach (var callBackObject in _callBacks)
```
needs a lock

moozzyk commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +440,51 @@ private void SetTransferMode(TransferMode transferMode)
+        {
+            public int CallBackID { get; }
+            private Func<byte[], object, Task> CallBack { get; set; }
+            private object State { get; set; }
```
These are private - why they need to be properties instead of readonly fields?

moozzyk commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +440,51 @@ private void SetTransferMode(TransferMode transferMode)
+            }
+            public void Dispose()
+            {
+                _handlers.RemoveAll((callBack) => { return callBack.CallBackID == _callBackId; });
```
use just reference to remove  (and also needs a lock)

moozzyk commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +440,51 @@ private void SetTransferMode(TransferMode transferMode)
+            {
+                _callBackId = callBackId;
+                _handlers = handler;
+            }
```
nit: separate with empty line

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.cs:
```diff
@@ -289,16 +289,16 @@ public void CannotStartConnectionWithInvalidTransportType(TransportType requeste
-            var receivedInvoked = false;
-            connection.Received += m =>
+            var onReceivedInvoked = false;
+            connection.OnReceived((_, __) =>
```
I thought you could use `(_, _)`..

mikaelm12 replied:
Nope. They have to be different

moozzyk commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -338,17 +341,10 @@ private void SetTransferMode(TransferMode transferMode)
-                                {
-                                    _logger.ExceptionThrownFromHandler(_connectionId, nameof(Received), ex);
-                                }
+                                await callBackObject.Invoke(buffer);
```
What happens if it throws?

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.cs:
```diff
@@ -672,132 +669,6 @@ public void CannotStartConnectionWithInvalidTransportType(TransportType requeste
         }
 
         [Fact]
-        public async Task CanReceiveDataEvenIfExceptionThrownFromPreviousReceivedEvent()
```
I think these tests are still relevant - user code can throw and I don't see exceptions to be handled.

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.cs:
```diff
@@ -672,132 +669,6 @@ public void CannotStartConnectionWithInvalidTransportType(TransportType requeste
-        }
-
-        [Fact]
-        public async Task CanReceiveDataEvenIfExceptionThrownSynchronouslyFromPreviousReceivedEvent()
```
ditto

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/HubConnectionTests.cs:
```diff
@@ -120,23 +120,6 @@ public class HubConnectionTests
         }
 
-        [Fact]
-        public async Task DoesNotThrowWhenClientMethodCalledButNoInvocationHandlerHasBeenSetUp()
```
This test should not have been removed.

mikaelm12 replied:
Replaced this test with a functional test  https://github.com/aspnet/SignalR/blob/mikaelm12/ClientAPIPart2/test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/HubConnectionTests.cs#L169

davidfowl commented on samples/ClientSample/RawSample.cs:
```diff
@@ -40,7 +40,7 @@ internal static void Register(CommandLineApplication app)
             {
                 var cts = new CancellationTokenSource();
-                connection.Received += data => Console.Out.WriteLineAsync($"{Encoding.UTF8.GetString(data)}");
+                connection.OnReceived((data, _) => Console.Out.WriteLineAsync($"{Encoding.UTF8.GetString(data)}"));
```
This should be:

```
connection.OnReceived(data => Console.Out.WriteLineAsync($"{Encoding.UTF8.GetString(data)}"))
```

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -272,18 +284,27 @@ private Task Shutdown(Exception ex = null)
+                copiedHandlers = new InvocationHandler[handlers.Count];
+                handlers.CopyTo(copiedHandlers);
+            }
+            var tasks = new List<Task>();
```
Not needed.

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -35,7 +35,8 @@ public class HubConnection
         private readonly Dictionary<string, InvocationRequest> _pendingCalls = new Dictionary<string, InvocationRequest>();
-        private readonly ConcurrentDictionary<string, InvocationHandler> _handlers = new ConcurrentDictionary<string, InvocationHandler>();
+        private readonly ConcurrentDictionary<string, List<InvocationHandler>> _handlers = new ConcurrentDictionary<string, List<InvocationHandler>>();
+        private object _handlerLock = new object();
```
What is this for?

mikaelm12 replied:
For safely copying and removing callbacks in dispose

davidfowl replied:
The lock should be per list not per connection.

davidfowl replied:
Just use the list as the lock.

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -398,24 +440,32 @@ public Type GetReturnType(string invocationId)
+                {
+                    return handlers[0].ParameterTypes;
+                }
+                throw new FormatException($"There are no callbacks registered for the method '{methodName}'");
```
FormatException?

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnectionExtensions.cs:
```diff
@@ -10,69 +10,79 @@ namespace Microsoft.AspNetCore.SignalR.Client
         }
 
-        public static void On(this HubConnection hubConnection, string methodName, Action handler)
+        public static IDisposable On(this HubConnection hubConnetion, string methodName, Type[] parameterTypes, Func<object[], object, Task> handler)
```
This extension is silly. Change it to also take in `object state`
Also, add one that takes `Func<object[], Task> handler`

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnectionExtensions.cs:
```diff
@@ -10,69 +10,79 @@ namespace Microsoft.AspNetCore.SignalR.Client
+        {
+            return hubConnetion.On(methodName, parameterTypes, (parameters, _) =>
+            {
+                handler(parameters, null);
```
return the handler call, right now you fire and forget

BrennanConroy commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -37,15 +39,15 @@ public class HttpConnection : IConnection
         private ReadableChannel<byte[]> Input => _transportChannel.In;
         private WritableChannel<SendMessage> Output => _transportChannel.Out;
+        private List<ReceiveCallBack> _callBacks = new List<ReceiveCallBack>();
+        private object _callbackLock = new object();
```
I think you can just use the list as the lock, same as mentioned above.

BrennanConroy commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -338,17 +340,17 @@ private void SetTransferMode(TransferMode transferMode)
 
-                            var receivedHandler = Received;
-                            if (receivedHandler != null)
+                            //Copying the callbacks to avoid concurrency issues
```
nit: space

BrennanConroy commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +446,56 @@ private void SetTransferMode(TransferMode transferMode)
+        public IDisposable OnReceived(Func<byte[], object, Task> callback, object state)
+        {
+            var callBack = new ReceiveCallBack(callback, state);
+            _callBacks.Add(callBack);
```
lock needed

BrennanConroy commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +446,56 @@ private void SetTransferMode(TransferMode transferMode)
+            return new Subscription(callBack, _callBacks, _callbackLock);
+        }
+
+        public IDisposable OnReceived(Func<byte[], object, Task> callback)
```
This should be an extensions method probably

BrennanConroy commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +446,56 @@ private void SetTransferMode(TransferMode transferMode)
+                State = state;
+            }
+
+            public Task Invoke(byte[] data)
```
nit: `InvokeAsync`

BrennanConroy commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.cs:
```diff
@@ -289,16 +289,16 @@ public void CannotStartConnectionWithInvalidTransportType(TransportType requeste
+                onReceivedInvoked = true;
                 return Task.CompletedTask;
-            };
+            }, null);
```
super nit: pass `onReceivedInvoked` as the state
And for other tests

mikaelm12 replied:
Talked offline. Doesn't apply for this test

BrennanConroy commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.cs:
```diff
@@ -336,12 +336,11 @@ public void CannotStartConnectionWithInvalidTransportType(TransportType requeste
-            connection.Received +=
-                async m =>
+            connection.OnReceived(async (_, __) =>
                 {
```
format

BrennanConroy commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/TestConnection.cs:
```diff
@@ -30,14 +32,16 @@ internal class TestConnection : IConnection
         public ReadableChannel<byte[]> SentMessages => _sentMessages.In;
         public WritableChannel<byte[]> ReceivedMessages => _receivedMessages.Out;
 
+        private ConcurrentDictionary<int, Tuple<Func<byte[], object, Task>, object>> _callBacks = new ConcurrentDictionary<int, Tuple<Func<byte[], object, Task>, object>>();
```
two lines

BrennanConroy replied:
I think `KeyValuePair` is more appropriate

BrennanConroy commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/TestConnection.cs:
```diff
@@ -136,5 +143,30 @@ private byte[] FormatMessageToArray(byte[] message)
+
+        public IDisposable OnReceived(Func<byte[], object, Task> callback, object state)
+        {
+            var id = GetNextCallBackId(); ;
```
remove extra `;`

BrennanConroy commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/TestConnection.cs:
```diff
@@ -136,5 +143,30 @@ private byte[] FormatMessageToArray(byte[] message)
+            {
+                _callBackId = callBackId;
+                _handler = handler;
+            }
```
newline

mikaelm12 commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -62,7 +63,7 @@ public HubConnection(IConnection connection, IHubProtocol protocol, ILoggerFacto
             _loggerFactory = loggerFactory ?? NullLoggerFactory.Instance;
             _logger = _loggerFactory.CreateLogger<HubConnection>();
-            _connection.Received += OnDataReceivedAsync;
+            _connection.OnReceived((data, state) => OnDataReceivedAsync(data), this);
```
Got it 😄

mikaelm12 commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -119,10 +120,21 @@ private IDataEncoder GetDataEncoder(TransferMode requestedTransferMode, Transfer
-            var invocationHandler = new InvocationHandler(parameterTypes, handler);
-            _handlers.AddOrUpdate(methodName, invocationHandler, (_, __) => invocationHandler);
+            var invocationHandler = new InvocationHandler(parameterTypes, handler, state);
+            if (!_handlers.TryGetValue(methodName, out var callBackList))
```
I did it this way so that I would conditionally create the `List<InvocationHandler>`.  If I just call add or update I have to create a new list every time.

BrennanConroy replied:
`AddOrUpdate` has an overload for a factory for the add part, so you don't need to create a new list every time.

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -272,18 +283,26 @@ private Task Shutdown(Exception ex = null)
+            }
+            foreach (var handler in copiedHandlers)
+            {
+                await handler.InvokeAsync(invocation.Arguments, handler.State);
```
Should we aggregate exceptions here? @moozzyk

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -398,24 +436,37 @@ public Type GetReturnType(string invocationId)
             public Type[] GetParameterTypes(string methodName)
             {
-                if (!_connection._handlers.TryGetValue(methodName, out InvocationHandler handler))
+                if (!_connection._handlers.TryGetValue(methodName, out List<InvocationHandler> handlers))
```
`out var`

mikaelm12 commented:
Updated. Addressed everything but skipping the `DoesNotThrowWhenClientMethodCalledButNoInvocationHandlerHasBeenSetUp` test in HubConnectionTests atm 😅

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -119,10 +119,20 @@ private IDataEncoder GetDataEncoder(TransferMode requestedTransferMode, Transfer
+                {
+                    return new List<InvocationHandler> { invocationHandler };
+                }, 
+                (_, invocations) => {
```
`{` on new line

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -377,6 +395,25 @@ private bool TryRemoveInvocation(string invocationId, out InvocationRequest irq)
+            {
+                _handler = handler;
+                _handlerList = handlerList;
+            }
```
nit: newline

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -398,24 +435,40 @@ public Type GetReturnType(string invocationId)
+                State = state;
+            }
+
+            public async Task InvokeAsync(object[] parameters, object state)
```
You already have state stored, don't need to pass it in

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnectionExtensions.cs:
```diff
@@ -10,69 +10,78 @@ namespace Microsoft.AspNetCore.SignalR.Client
+            return hubConnetion.On(methodName, parameterTypes, (parameters, _) =>
+            {
+                return handler(parameters);
+            }, null);
```
pass handler as state

BrennanConroy commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -338,16 +339,23 @@ private void SetTransferMode(TransferMode transferMode)
                                 catch (Exception ex)
                                 {
-                                    _logger.ExceptionThrownFromHandler(_connectionId, nameof(Received), ex);
+                                    _logger.ExceptionThrownFromHandler(_connectionId, "Test", ex);
```
"Test"?

mikaelm12 replied:
oops

BrennanConroy commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +452,57 @@ private void SetTransferMode(TransferMode transferMode)
+            return OnReceived((data, _) => { return callback(data); }, null);
+        }
+
+        public class ReceiveCallBack
```
private

mikaelm12 replied:
So I'm also using these classes in TestConnection. I'll just copy them then

BrennanConroy commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +452,57 @@ private void SetTransferMode(TransferMode transferMode)
+            }
+        }
+
+        public class Subscription : IDisposable
```
private

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -398,24 +437,40 @@ public Type GetReturnType(string invocationId)
-            public Func<object[], Task> Handler { get; }
+            public Func<object[], object, Task> Callback { get; }
             public Type[] ParameterTypes { get; }
+            public object State { get; }
```
nit: `private readonly`

BrennanConroy commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -338,16 +339,23 @@ private void SetTransferMode(TransferMode transferMode)
                                 catch (Exception ex)
                                 {
-                                    _logger.ExceptionThrownFromHandler(_connectionId, nameof(Received), ex);
+                                    _logger.ExceptionThrownFromHandler(_connectionId, nameof(callBackObject), ex);
```
Log message was about the 'name' event handler, something needs to change here.

BrennanConroy replied:
Message is now "An exception was thrown from the 'callBackObject' callback". I think 'OnReceived' is more appropriate. And the function name should be `ExceptionThrownFromCallback`

BrennanConroy commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +452,57 @@ private void SetTransferMode(TransferMode transferMode)
+
+        public IDisposable OnReceived(Func<byte[], Task> callback)
+        {
+            return OnReceived((data, _) => { return callback(data); }, null);
```
`callback` can be passed as state

BrennanConroy commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +452,57 @@ private void SetTransferMode(TransferMode transferMode)
+
+        private class ReceiveCallBack
+        {
+            public readonly Func<byte[], object, Task> CallBack;
```
nit: private

moozzyk commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -119,10 +119,21 @@ private IDataEncoder GetDataEncoder(TransferMode requestedTransferMode, Transfer
-            _handlers.AddOrUpdate(methodName, invocationHandler, (_, __) => invocationHandler);
+            var invocationHandler = new InvocationHandler(parameterTypes, handler, state);
+            var invocationList = _handlers.AddOrUpdate(methodName, 
+                _ => 
```
nit: could be just `_ => new List<InvocationHandler>{ invocationHandler }`

moozzyk commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -398,24 +437,40 @@ public Type GetReturnType(string invocationId)
-            public Func<object[], Task> Handler { get; }
+            public Func<object[], object, Task> Callback { get; }
             public Type[] ParameterTypes { get; }
+            private readonly object State;
```
nit: should be `_state`

moozzyk commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -398,24 +437,40 @@ public Type GetReturnType(string invocationId)
         private struct InvocationHandler
         {
-            public Func<object[], Task> Handler { get; }
+            public Func<object[], object, Task> Callback { get; }
```
I think this should not be public.

moozzyk commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -37,15 +39,14 @@ public class HttpConnection : IConnection
-
         private ReadableChannel<byte[]> Input => _transportChannel.In;
         private WritableChannel<SendMessage> Output => _transportChannel.Out;
+        private List<ReceiveCallBack> _callBacks = new List<ReceiveCallBack>();
```
nit: casing - `_callbacks`  `ReceiveCallback` etc. (including below)

davidfowl replied:
readonly

moozzyk commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +452,61 @@ private void SetTransferMode(TransferMode transferMode)
+            return new Subscription(callBack, _callBacks);
+        }
+
+        public IDisposable OnReceived(Func<byte[], Task> callback)
```
I don't think this should be on HttpConnection - this is just syntactic sugar.

mikaelm12 replied:
Should I make a extensions file for just this method?

moozzyk replied:
Yes.

moozzyk commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +452,61 @@ private void SetTransferMode(TransferMode transferMode)
+        private class ReceiveCallBack
+        {
+            private readonly Func<byte[], object, Task> CallBack;
+            private readonly object State;
```
naming - `_state`, `_callback` - these are class variables

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.cs:
```diff
@@ -762,17 +760,17 @@ public void CannotStartConnectionWithInvalidTransportType(TransportType requeste
-                    return Task.CompletedTask;
-                };
+                connection.OnReceived((data) =>
+               {
```
formatting

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/HubConnectionTests.cs:
```diff
@@ -120,7 +120,7 @@ public class HubConnectionTests
         }
 
-        [Fact]
+        [Fact(Skip = "Will add this back")]
```
Fix the test, remove `Skip`

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/TestConnection.cs:
```diff
@@ -13,6 +16,7 @@
 using Microsoft.AspNetCore.Sockets.Client;
 using Microsoft.AspNetCore.Sockets.Features;
 using Newtonsoft.Json;
+using static Microsoft.AspNetCore.Sockets.Client.HttpConnection;
```
needed?

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/TestConnection.cs:
```diff
@@ -2,6 +2,9 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Concurrent;
```
at least some of these don't seem to be needed

moozzyk commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnectionExtensions.cs:
```diff
@@ -10,69 +10,79 @@ namespace Microsoft.AspNetCore.SignalR.Client
         }
 
-        public static void On(this HubConnection hubConnection, string methodName, Action handler)
+        public static IDisposable On(this HubConnection hubConnetion, string methodName, Type[] parameterTypes, Func<object[], Task> handler)
```
Can you move this to the bottom of this file (i.e. as the last method)? This separates closely related extension methods.

moozzyk commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnectionExtensions.cs:
```diff
@@ -10,69 +10,79 @@ namespace Microsoft.AspNetCore.SignalR.Client
-        public static void On(this HubConnection hubConnection, string methodName, Action handler)
+        public static IDisposable On(this HubConnection hubConnetion, string methodName, Type[] parameterTypes, Func<object[], Task> handler)
+        {
+            return hubConnetion.On(methodName, parameterTypes, (parameters, state) =>
```
typo: should be `hubConnection`

moozzyk commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnectionExtensions.cs:
```diff
@@ -10,69 +10,79 @@ namespace Microsoft.AspNetCore.SignalR.Client
         }
 
-        public static void On(this HubConnection hubConnection, string methodName, Action handler)
+        public static IDisposable On(this HubConnection hubConnetion, string methodName, Type[] parameterTypes, Func<object[], Task> handler)
```
Should we actually have overloads for different parameters instead in addition to this - similarly to what we have for `Actions`?

mikaelm12 replied:
Hmm. Maybe I can add this in another PR?

davidfowl commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -272,18 +279,26 @@ private Task Shutdown(Exception ex = null)
-            // TODO: Dispatch to a sync context to ensure we aren't blocking this loop.
-            return handler.Handler(invocation.Arguments);
+            // Copying the callbacks to avoid concurrency issues
+            InvocationHandler[] copiedHandlers;
```
Put a todo here to optimize this

BrennanConroy commented on test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/HubConnectionTests.cs:
```diff
@@ -166,6 +166,37 @@ public HubConnectionTests(ServerFixture<Startup> serverFixture, ITestOutputHelpe
+                }
+                finally
+                {
+                    Assert.False(exceptionThrown);
```
Remove

BrennanConroy commented on test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/HubConnectionTests.cs:
```diff
@@ -166,6 +166,37 @@ public HubConnectionTests(ServerFixture<Startup> serverFixture, ITestOutputHelpe
+                }
+                catch (Exception ex)
+                {
+                    exceptionThrown = true;
```
not needed

BrennanConroy commented on test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/Hubs.cs:
```diff
@@ -24,7 +24,15 @@ public string Echo(string message)
+            await Clients.Client(Context.ConnectionId).InvokeAsync("NoClientHandler");
+        }
+
+        public void NoClientHandler() { }
```
These shouldn't be needed right?

BrennanConroy commented on test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/Hubs.cs:
```diff
@@ -24,7 +24,15 @@ public string Echo(string message)
+        public void NoClientHandler() { }
+
         public IObservable<int> Stream(int count)
+
```
revert

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -425,24 +461,40 @@ public Type GetReturnType(string invocationId)
+
+            public async Task InvokeAsync(object[] parameters)
+            {
+                await _callback(parameters, _state);
```
nit: just return, no need to await

BrennanConroy commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +452,52 @@ private void SetTransferMode(TransferMode transferMode)
+            return new Subscription(callBack, _callbacks);
+        }
+
+        private class ReceiveCallBack
```
nit: ReceiveCallback

BrennanConroy commented on src/Microsoft.AspNetCore.Sockets.Client.Http/Internal/SocketClientLoggerExtensions.cs:
```diff
@@ -151,7 +151,7 @@ internal static class SocketClientLoggerExtensions
 
         private static readonly Action<ILogger, DateTime, string, string, Exception> _exceptionThrownFromHandler =
-            LoggerMessage.Define<DateTime, string, string>(LogLevel.Error, new EventId(19, nameof(ExceptionThrownFromHandler)), "{time}: Connection Id {connectionId}: An exception was thrown from the '{eventHandlerName}' event handler.");
+            LoggerMessage.Define<DateTime, string, string>(LogLevel.Error, new EventId(19, nameof(ExceptionThrownFromHandler)), "{time}: Connection Id {connectionId}: An exception was thrown from the '{eventHandlerName}' callback");
```
`{eventHandlerName}` will need to be changed also

mikaelm12 commented:
Updated

BrennanConroy commented on src/Microsoft.AspNetCore.SignalR.Client.Core/Internal/SignalRClientLoggerExtensions.cs:
```diff
@@ -103,6 +103,10 @@ internal static class SignalRClientLoggerExtensions
+        private static readonly Action<ILogger, string, Exception> _exceptionThrownFromCallback =
+           LoggerMessage.Define<string>(LogLevel.Error, new EventId(23, nameof(ExceptionThrownFromCallback)), "An exception was thrown from the '{callback}' callback");
+
+
```
nit: remove newline

moozzyk commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -299,18 +306,34 @@ private Task Shutdown(Exception ex = null)
+            {
+                copiedHandlers = new InvocationHandler[handlers.Count];
+                handlers.CopyTo(copiedHandlers);
+            }
```
nit: add a new line

moozzyk commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -119,10 +119,17 @@ private IDataEncoder GetDataEncoder(TransferMode requestedTransferMode, Transfer
+            var invocationList = _handlers.AddOrUpdate(methodName,  _ =>  new List<InvocationHandler> { invocationHandler }, 
+                (_, invocations) =>
+                {
+                    invocations.Add(invocationHandler);
```
doesn't this need a lock? You are updating the a which can be used elsewhere.

moozzyk commented on src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:
```diff
@@ -425,24 +468,40 @@ public Type GetReturnType(string invocationId)
         private struct InvocationHandler
         {
-            public Func<object[], Task> Handler { get; }
+            private Func<object[], object, Task> _callback;
```
nit: _callback should be `readonly` 
also group private variables together

moozzyk commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -37,15 +39,14 @@ public class HttpConnection : IConnection
-
         private ReadableChannel<byte[]> Input => _transportChannel.In;
         private WritableChannel<SendMessage> Output => _transportChannel.Out;
+        private List<ReceiveCallback> _callbacks = new List<ReceiveCallback>();
```
`readonly`

moozzyk commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +452,52 @@ private void SetTransferMode(TransferMode transferMode)
 
+        public IDisposable OnReceived(Func<byte[], object, Task> callback, object state)
+        {
+            var callBack = new ReceiveCallback(callback, state);
```
you need to rename this variable - you have two different variables which differ just by casing in one place. Call it `receiveCallback` or something to that effect.

moozzyk commented on src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs:
```diff
@@ -444,6 +452,52 @@ private void SetTransferMode(TransferMode transferMode)
+            private readonly Func<byte[], object, Task> _callback;
+            private readonly object _state;
+
+            public ReceiveCallback(Func<byte[], object, Task> callBack, object state)
```
`callBack` fix casing

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/HubConnectionTests.cs:
```diff
@@ -166,6 +166,33 @@ public HubConnectionTests(ServerFixture<Startup> serverFixture, ITestOutputHelpe
+                    await connection.StartAsync().OrTimeout();
+
+                    await connection.InvokeAsync("CallHandlerThatDoesntExist").OrTimeout();
+
```
nit: add empty line
I think you may also want to check that there is no exception in the close event. I think if the connection was closed due to an exception it would show up there.

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.cs:
```diff
@@ -139,7 +139,7 @@ public void CannotStartConnectionWithInvalidTransportType(TransportType requeste
             var transport = new Mock<ITransport>();
             transport.Setup(t => t.StopAsync()).Returns(async () => { await releaseDisposeTcs.Task; });
-            var connection = new HttpConnection(new Uri("http://fakeuri.org/"), new TestTransportFactory(transport.Object), loggerFactory: null, httpMessageHandler: mockHttpHandler.Object);
+            var connection = new HttpConnection(new Uri("http://fakeuri.org/"), transportFactory: new TestTransportFactory(transport.Object), loggerFactory: null, httpMessageHandler: mockHttpHandler.Object);
```
I don't think you need the name - the type we new up tells you what you need to know. (I don't care too much though)

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.cs:
```diff
@@ -255,7 +255,7 @@ public void CannotStartConnectionWithInvalidTransportType(TransportType requeste
 
         [Fact]
-        public async Task ReceivedEventNotRaisedAfterConnectionIsDisposed()
+        public async Task ReceivedCallBackNotRaisedAfterConnectionIsDisposed()
```
nit: Callback casing.

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.cs:
```diff
@@ -336,12 +336,11 @@ public void CannotStartConnectionWithInvalidTransportType(TransportType requeste
+            connection.OnReceived(async _ =>
                 {
                     callbackInvokedTcs.SetResult(null);
                     await closedTcs.Task;
```
you could just return the task I think.

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.cs:
```diff
@@ -392,11 +391,10 @@ public void CannotStartConnectionWithInvalidTransportType(TransportType requeste
             var connection = new HttpConnection(new Uri("http://fakeuri.org/"), new TestTransportFactory(mockTransport.Object), loggerFactory: null, httpMessageHandler: mockHttpHandler.Object);
-            connection.Received +=
-                async m =>
+            connection.OnReceived(async _ =>
```
`connection.OnReceived(_ => blockReceiveCallbackTcs.Task);`

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.cs:
```diff
@@ -762,17 +760,17 @@ public void CannotStartConnectionWithInvalidTransportType(TransportType requeste
-                        receivedRaised = true;
-                        throw new InvalidOperationException();
-                    }
+                   if (!receivedRaised)
```
nit: indentation is off

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/TestConnection.cs:
```diff
@@ -30,14 +31,15 @@ internal class TestConnection : IConnection
         public ReadableChannel<byte[]> SentMessages => _sentMessages.In;
         public WritableChannel<byte[]> ReceivedMessages => _receivedMessages.Out;
 
+        private List<ReceiveCallBack> _callbacks = new List<ReceiveCallBack>();
```
`readonly`

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/TestConnection.cs:
```diff
@@ -136,5 +141,48 @@ private byte[] FormatMessageToArray(byte[] message)
+            return new Subscription(receiveCallBack, _callbacks);
+        }
+
+        private class ReceiveCallBack
```
casing

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/TestConnection.cs:
```diff
@@ -136,5 +141,48 @@ private byte[] FormatMessageToArray(byte[] message)
+
+        private class ReceiveCallBack
+        {
+            public readonly Func<byte[], object, Task> CallBack;
```
make it a getter (and fix casing)

mikaelm12 replied:
This should actually be private since I've added the InvokeAsync method

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/TestConnection.cs:
```diff
@@ -136,5 +141,48 @@ private byte[] FormatMessageToArray(byte[] message)
+
+            public void Dispose()
+            {
+                lock (_callbacks)
```
locking in just one place does not make much sense. I was thinking we should make the callback list threadsafe but since we don't add/remove callbacks in our tests after the connection is started we should be fine. We might need to do this in the future if we start seeing errors caused by this.

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Tests/EndToEndTests.cs:
```diff
@@ -224,12 +225,12 @@ public static IEnumerable<object[]> MessageSizesData
                         receiveTcs.TrySetResult(data);
                         return Task.CompletedTask;
-                    };
+                    }, null);
```
you could pass `receiveTcs` as `state`

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Tests/EndToEndTests.cs:
```diff
@@ -127,7 +127,8 @@ public EndToEndTests(ServerFixture<Startup> serverFixture, ITestOutputHelper out
                         receiveTcs.TrySetResult(Encoding.UTF8.GetString(data));
                         return Task.CompletedTask;
-                    };
+                    }, null);
```
you could pass `receiveTcs` as `state`

moozzyk commented on test/Microsoft.AspNetCore.SignalR.Client.Tests/TestConnection.cs:
```diff
@@ -122,7 +122,14 @@ private byte[] FormatMessageToArray(byte[] message)
+                            lock (_callbacks)
+                            {
+                                callbackCopies = new ReceiveCallback[_callbacks.Count];
+                                _callbacks.CopyTo(callbackCopies);
```
Can you just do `_callbacks.ToArray()`?
